### PR TITLE
Geopandas OGC API Features (OAF) Support (USGS/WMA - 33)

### PR DESCRIPTION
**Pull request by C-Loftus**
- Adds support for Geopandas in pygeoapi
- Supports propertyfilters/display resulttype bbox datetime sortby skipGeometry for geopandas.
- Adds unit tests for Geopandas provider
- Adds data for unit tests
- Passes existing CSV tests as well as gpkg tests for https://reference.geoconnex.us/collections/hu02/items

Note: I tested this all inside the current pygeoapi master inside a devcontainer docker volume. So I have not tested this in the CI pipeline for this repo yet. Will iterate if there are failures.

Only thing I am currently unsure on:
- Sometimes the datatypes of the file read in by the geopandas `read_file` don't match the proper datatype. I am not entirely sure how to fix this in a generalized way since the file we are reading in doesn't always specify the type of the data (i.e. CSV) and might require a heuristic, but sometimes the geopandas heuristic is incorrect. The only place this matters is `get_fields` since it returns the type associated with the field.

### CODE REVIEW

**Comment by C-Loftus** [comment]
Appears that the sitemap generator is broken.

File: README.md [hunk 1]
```diff
@@ -78,6 +79,16 @@ The SPARQL Provider only uses variables prefixed with sparql\_ in the configurat
 - `sparql_subject`: The SPARQL variable representing the subject URI in the query.
 - `sparql_predicates`: A mapping of attribute names to SPARQL predicates. These predicates will be used to query specific attributes in the SPARQL data source.
 
+### GeoPandas
+
+The GeoPandas Provider enables OGC API - Feature support using GeoPandas as the backend. This integration can read in data files in [any of the geospatial formats supported by GeoPandas](https://geopandas.org/en/stable/docs/user_guide/io.html#supported-drivers-file-formats).
+
+`id_field` and `geometry` are the only two special columns required.
```
Comment (webb-ben) [inline README.md:86]: can you add yaml block representations here

File: README.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ Additional OGC API - Feature providers are listed below, along with a matrix of
 | `CKAN`             | ✅/✅                    | results/hits | ❌   | ❌       | ✅      | ✅            | ❌  | ❌           | ✅  |
 | `PsuedoPostgreSQL` | ✅/✅                    | results/hits | ✅   | ✅       | ✅      | ✅            | ✅  | ❌           | ✅  |
 | `SPARQL`           | ❌/✅                    | results/hits | ❌   | ❌       | ❌      | ❌            | ❌  | ❌           | ❌  |
+| `GeoPandas`        | ✅/✅                    | results/hits | ✅   | ✅       | ✅      | ✅            | ❌  | ❌           | ❌  |
```
Comment (webb-ben) [inline README.md:14]: It looks like you added transactions? And you should add CRS (which is just a simple decorator function (see https://github.com/geopython/pygeoapi/pull/1729/files)

File: pygeoapi_plugins/provider/geopandas_provider.py
Comment (webb-ben) [inline pygeoapi_plugins/provider/geopandas_provider.py:1]: rename to `geopandas_.py`

File: pygeoapi_plugins/provider/geopandas_provider.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+from collections import OrderedDict
+import datetime
+import geopandas
+import pandas
+from shapely import box
+import shapely.geometry
+import logging
+from typing import Literal, Optional
+from typing import TypedDict
+from collections import defaultdict
```
Comment (webb-ben) [inline pygeoapi_plugins/provider/geopandas_provider.py:39]: Move to import above?

File: pygeoapi_plugins/provider/geopandas_provider.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        
+        our_types_names = {k: {
+            "type": pandas_default[v]
+            }
+        for k, v in field_mapper.items()
+        }
+
+        return our_types_names
+
+    def get_schema(self, schema_type: SchemaType = SchemaType.item):
```
Comment (webb-ben) [inline pygeoapi_plugins/provider/geopandas_provider.py:245]: No need to implement unimplemented methods of the Base Provider. For all of the methods returning `NotImplementedEror()`s

File: pygeoapi_plugins/provider/geopandas_provider.py [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+        Provide data/file metadata
+
+        :returns: `dict` of metadata construct (format
+                  determined by provider/standard)
+        """
+
+        raise NotImplementedError()
+
+    # we want to support bbox and datetime_
+    def query(self, offset=0, limit=10, resulttype: Literal['results', 'hits']='results',
```
Comment (webb-ben) [inline pygeoapi_plugins/provider/geopandas_provider.py:282]: add **kwargs to arguments of this method. https://github.com/geopython/pygeoapi/blob/master/pygeoapi/provider/sensorthings.py#L173. Throws an error if you try to use the provider in pygeoapi

File: pygeoapi_plugins/provider/geopandas_provider.py [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+            if skip_geometry:
+                feature['geometry'] = None
+            else:
+                if hasattr(self, 'geometry_x') and hasattr(self, 'geometry_y'):
+                    feature['geometry']["coordinates"] = [float(row[self.geometry_x]),
+                                                          float(row[self.geometry_y])]
+                    feature['geometry']['type'] = 'Point'
+                elif hasattr(self, 'geometry_col'):
+                    feature['geometry']["coordinates"] = row[self.geometry_col]
```
Comment (webb-ben) [inline pygeoapi_plugins/provider/geopandas_provider.py:378]: You will need to cast the geometry here to geojson

File: tests/test_geopandas_provider.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+
+@pytest.fixture()
+def gpkg_config():
+    return {
+        'name': 'gpkg',
+        'type': 'feature',
+        'data': "data/hu02.gpkg",
+        'id_field': 'HUC2',
+    }
```
Comment (webb-ben) [inline tests/test_geopandas_provider.py:160]: Declare all the configs at the top of the file. You should also be able to use the remote URL for this file.

```python
    'data': 'https://www.hydroshare.org/resource/3295a17b4cc24d34bd6a5c5aaf753c50/data/contents/hu02.gpkg'
```

Could reduce the number of files needed to run the tests by referencing stable remote sources.

**Review by webb-ben — CHANGES_REQUESTED** [review]
Overall looks good! 

Some small changes. Can you also use consistent quoting in your python files, I prefer single quotes. You will also need to do some formatting to pass flake8 tests.

File: pygeoapi_plugins/provider/geopandas_provider.py (hunk 3 above)
Comment (C-Loftus) [reply pygeoapi_plugins/provider/geopandas_provider.py:39]: Sorry what do you mean by this?

**Comment by C-Loftus** [comment]
> Overall looks good!
> 
> Some small changes. Can you also use consistent quoting in your python files, I prefer single quotes. You will also need to do some formatting to pass flake8 tests.

Sounds good. Working on changes. By the way, what is your opinion on an autoformatter pre-commit for this repo? Wasn't sure if you want to keep this the same as the pygeoapi style. I honestly don't think about quote style or import order much since I usually just delegate that to ruff.

But obviously happy to change things however you prefer.

File: tests/test_geopandas_provider.py (hunk 7 above)
Comment (C-Loftus) [reply tests/test_geopandas_provider.py:160]: Can we keep this locally? The dataset is not super large but at the same time it significantly slows down test driven development to need to fetch it every time the environment is spun up. In general I also think that unit tests should isolate environment predictably and not fetch external resources. Even if the resource is stable I think it is stablest to have it in the repo and then remove it during packaging if we don't want to ship with it.

I can't change the pypi code to remove it before publishing

**Comment by webb-ben** [comment]
Why don't you add precognition here as well then? Fine by me

**Review by webb-ben — APPROVED** [review]

**Pull request merged**